File: services/client/src/messages/messages.go
package messages

import (
	"encoding/binary"
	"errors"
	"io"
	"strconv"
	"strings"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/logger"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

type MessageType uint8

const (
	ITEMS_PER_DATE      = 3
	ITEMS_PER_LINE      = 5
	MAX_NAME_LEN        = 50
	UINT                = 64
	BODY_MIN_SIZE       = 12 // en bytes
	MESSAGE_HEADER_SIZE = 11 // en bytes
	BATCH_HEADER_SIZE   = 3
)

const (
	NameField = iota
	SurnameField
	DniField
	DateField
	BetField
)

const (
	CONNECT         MessageType = 1   // 00000001
	CONNECT_ACK     MessageType = 129 // 10000001
	BATCH           MessageType = 16  // 00010000
	BATCH_ACK       MessageType = 144 // 10010000
	BET             MessageType = 64  // 01000000
	BET_ACK         MessageType = 192 // 11000000
	BET_END         MessageType = 66  // 01000010
	WINNER          MessageType = 32  // 00100000
	WINNER_ACK      MessageType = 160 // 10100000
	CONNECT_END     MessageType = 3   // 00000011
	CONNECT_END_ACK MessageType = 131 // 10000011
	ERROR           MessageType = 255 // 11111111
)

type Batch struct {
	Header   HeaderBatch
	Messages []Message
}

type HeaderBatch struct {
	Type      MessageType
	SizeBatch uint16
}

type HeaderMessage struct {
	Type        MessageType
	SeqNum      uint8
	AckNum      uint8
	AgencyId    uint32
	SizePayload uint16
	SizeName    uint8
	SizeSurName uint8
}

type BodyMessage struct {
	Dni     uint32
	Bet     uint32
	Year    uint16
	Month   uint8
	Day     uint8
	Name    string // max 50 bytes
	SurName string // max 50 bytes
}

type Message struct {
	Header HeaderMessage
	Body   *BodyMessage
}

func Read(socket io.Reader) ([]Message, error) {
	header_data, err := safe_socket.RecvAll(socket, BATCH_HEADER_SIZE)
	if err != nil {
		logger.Warn("read-message-header", logger.Fail)
		return nil, err
	}
	header, err := deserialize_batch_header(header_data)
	if err != nil {
		logger.Warn("deserialize-message-header", logger.Fail)
		return nil, err
	}

	var messages []Message

	for i := 0; i < int(header.SizeBatch); i++ {
		message, err := read_message(socket)
		if err != nil {
			logger.Warn("reading-messages", logger.Fail)
			return nil, err
		}
		messages = append(messages, message)
	}

	return messages, nil

}

func Build_batch(messages []Message) Batch {
	return Batch{
		Header: HeaderBatch{
			Type:      BATCH,
			SizeBatch: uint16(len(messages)),
		},
		Messages: messages,
	}
}

func Build_batch_buffer(batch_size int) []byte {
	size_message := MESSAGE_HEADER_SIZE + BODY_MIN_SIZE + MAX_NAME_LEN*2
	size_batch := BATCH_HEADER_SIZE + batch_size*size_message
	return make([]byte, BATCH_HEADER_SIZE, size_batch)
}

func Set_batch_size(data []byte, batch_size uint16) {
	data[0] = byte(BATCH)
	binary.BigEndian.PutUint16(data[1:3], batch_size)
}

func Append_bet(data []byte, seq uint8, id uint32, info_body string) ([]byte, error) {
	body, err := build_body(info_body)
	if err != nil {
		return data, err
	}

	header := build_header(BET, seq, 0, id, &body)
	message := Message{
		Header: header,
		Body:   &body,
	}

	return serialize_into(data, message), nil
}

func Serialize_batch(batch Batch) []byte {
	size_batch := BATCH_HEADER_SIZE

	for _, message := range batch.Messages {
		size_batch += serialized_message_size(message)
	}

	data := make([]byte, BATCH_HEADER_SIZE, size_batch)

	data[0] = byte(batch.Header.Type)
	binary.BigEndian.PutUint16(data[1:3], batch.Header.SizeBatch)

	for _, message := range batch.Messages {
		data = serialize_into(data, message)
	}

	return data
}

func deserialize_batch_header(data []byte) (HeaderBatch, error) {

	if len(data) < BATCH_HEADER_SIZE {
		logger.Warn("deserialize-message-header", logger.Fail)
		return HeaderBatch{}, errors.New("message is smaller than header")
	}

	header := HeaderBatch{
		Type:      MessageType(data[0]),
		SizeBatch: binary.BigEndian.Uint16(data[1:3]),
	}
	return header, nil
}

func read_message(socket io.Reader) (Message, error) {
	header_data, err := safe_socket.RecvAll(socket, MESSAGE_HEADER_SIZE)
	if err != nil {
		logger.Warn("read-message-header", logger.Fail)
		return Message{}, err
	}
	header, err := deserialize_header(header_data)
	if err != nil {
		logger.Warn("deserialize-message-header", logger.Fail)
		return Message{}, err
	}

	var body *BodyMessage

	if header.SizePayload > 0 {
		body_data, err := safe_socket.RecvAll(socket, int(header.SizePayload))
		if err != nil {
			logger.Warn("read-message-body", logger.Fail)
			return Message{}, err
		}
		decoded, err := deserialize_body(body_data, header)
		if err != nil {
			logger.Warn("deserialize-message-body", logger.Fail)
			return Message{}, err
		}
		body = &decoded
	}

	return Message{
		Header: header,
		Body:   body,
	}, nil

}

func Parse_string_uint(text string) (uint64, error) {
	value, err := strconv.ParseUint(text, 10, UINT)
	if err != nil {
		logger.Warn("parse-string-uint64", logger.Fail, text)
		return 0, errors.New("Error parsing string to uint")
	}
	return value, nil
}

func truncateUTF8Bytes(name string, maxBytes int) string {
	if len(name) <= maxBytes {
		return name
	}

	end := 0
	for i := range name {
		if i > maxBytes {
			break
		}
		end = i
	}
	return name[:end]
}

func Build_message(kind MessageType, seq uint8, ack uint8, id uint32, info_body string) (Message, error) {
	var (
		body       *BodyMessage
		header     HeaderMessage
		parsedBody BodyMessage
		err        error
	)

	switch kind {
	case BET:
		parsedBody, err = build_body(info_body)
		if err != nil {
			return Message{}, err
		}
		body = &parsedBody
	default:
		body = nil
	}

	header = build_header(kind, seq, ack, id, body)

	message := Message{
		Header: header,
		Body:   body,
	}

	return message, nil
}

func build_header(kind MessageType, seq uint8, ack uint8, id uint32, body *BodyMessage) HeaderMessage {
	var (
		sizePayload uint16
		sizeName    uint8
		sizeSurname uint8
	)

	if body != nil {
		sizeName = uint8(len(body.Name))
		sizeSurname = uint8(len(body.SurName))
		sizePayload = BODY_MIN_SIZE + uint16(sizeName) + uint16(sizeSurname)
	}
	header := HeaderMessage{
		Type:        kind,
		SeqNum:      seq,
		AckNum:      ack,
		AgencyId:    id,
		SizePayload: sizePayload,
		SizeName:    sizeName,
		SizeSurName: sizeSurname,
	}
	return header
}

func build_body(line string) (BodyMessage, error) {
	fields := strings.Split(line, ",")
	if len(fields) != ITEMS_PER_LINE {
		logger.Warn("serialize-line", logger.Fail, line)
		return BodyMessage{}, errors.New("Not enough items per line")
	}

	var (
		name    string
		surname string
		dni     uint32
		year    uint16
		month   uint8
		day     uint8
		bet     uint32
		errs    []error
	)

	for index, field := range fields {
		switch index {
		case NameField:
			name = truncateUTF8Bytes(field, MAX_NAME_LEN)

		case SurnameField:
			surname = truncateUTF8Bytes(field, MAX_NAME_LEN)

		case DniField:
			value, err := Parse_string_uint(field)
			if err != nil {
				logger.Warn("parse-dni-uint64", logger.Fail, value)
				errs = append(errs, errors.New("error parsing dni"))
			} else {
				dni = uint32(value)
			}

		case DateField:
			date := strings.Split(field, "-")
			if len(date) != ITEMS_PER_DATE {
				logger.Warn("split-date", logger.Fail, field)
				errs = append(errs, errors.New("error splitting date with -"))
				break
			}

			value, err := Parse_string_uint(date[0])
			if err != nil {
				logger.Warn("parse-year", logger.Fail, date[0])
				errs = append(errs, errors.New("error parsing year"))
			} else {
				year = uint16(value)
			}

			value, err = Parse_string_uint(date[1])
			if err != nil {
				logger.Warn("parse-month", logger.Fail, date[1])
				errs = append(errs, errors.New("error parsing month"))
			} else {
				month = uint8(value)
			}

			value, err = Parse_string_uint(date[2])
			if err != nil {
				logger.Warn("parse-month", logger.Fail, date[2])
				errs = append(errs, errors.New("error parsing day"))
			} else {
				day = uint8(value)
			}

		case BetField:
			value, err := Parse_string_uint(field)
			if err != nil {
				logger.Warn("parse-bet", logger.Fail, field)
				errs = append(errs, errors.New("error parsing bet amount"))
			} else {
				bet = uint32(value)
			}
		}
	}

	if len(errs) > 0 {
		return BodyMessage{}, errors.Join(errs...)
	}

	body := BodyMessage{
		Dni:     dni,
		Bet:     bet,
		Year:    year,
		Month:   month,
		Day:     day,
		Name:    name,
		SurName: surname,
	}

	return body, nil
}

func serialized_message_size(message Message) int {
	size_message := MESSAGE_HEADER_SIZE

	if message.Body != nil {
		size_message += int(message.Header.SizePayload)
	}

	return size_message
}

func serialize_into(data []byte, message Message) []byte {
	header := message.Header
	offset := len(data)
	data = data[:offset+serialized_message_size(message)]
	serialized_message := data[offset:]
	clear(serialized_message)

	serialized_message[0] = byte(header.Type)
	serialized_message[1] = header.SeqNum
	serialized_message[2] = header.AckNum
	binary.BigEndian.PutUint32(serialized_message[3:7], header.AgencyId)
	binary.BigEndian.PutUint16(serialized_message[7:9], header.SizePayload)
	serialized_message[9] = header.SizeName
	serialized_message[10] = header.SizeSurName

	if message.Body != nil {
		body := message.Body
		body_offset := MESSAGE_HEADER_SIZE

		binary.BigEndian.PutUint32(serialized_message[body_offset:body_offset+4], body.Dni)
		body_offset += 4
		binary.BigEndian.PutUint32(serialized_message[body_offset:body_offset+4], body.Bet)
		body_offset += 4
		binary.BigEndian.PutUint16(serialized_message[body_offset:body_offset+2], body.Year)
		body_offset += 2
		serialized_message[body_offset] = body.Month
		body_offset++
		serialized_message[body_offset] = body.Day
		body_offset++

		body_offset += copy(serialized_message[body_offset:], body.Name)
		copy(serialized_message[body_offset:], body.SurName)
	}

	return data
}

func Serialize(message Message) []byte {
	data := make([]byte, 0, serialized_message_size(message))
	return serialize_into(data, message)
}

func deserialize_header(data []byte) (HeaderMessage, error) {

	if len(data) < MESSAGE_HEADER_SIZE {
		logger.Warn("deserialize-message-header", logger.Fail)
		return HeaderMessage{}, errors.New("message is smaller than header")
	}

	header := HeaderMessage{
		Type:        MessageType(data[0]),
		SeqNum:      data[1],
		AckNum:      data[2],
		AgencyId:    binary.BigEndian.Uint32(data[3:7]),
		SizePayload: binary.BigEndian.Uint16(data[7:9]),
		SizeName:    data[9],
		SizeSurName: data[10],
	}
	return header, nil
}

func deserialize_body(data []byte, header HeaderMessage) (BodyMessage, error) {
	payloadSize := int(header.SizePayload)
	nameSize := int(header.SizeName)
	surnameSize := int(header.SizeSurName)

	if payloadSize < BODY_MIN_SIZE {
		logger.Warn("deserialize-message-payload", logger.Fail)
		return BodyMessage{}, errors.New("invalid payload size")
	}

	if payloadSize != BODY_MIN_SIZE+nameSize+surnameSize {
		logger.Warn("deserialize-message-payload", logger.Fail)
		return BodyMessage{}, errors.New("payload size does not match name sizes")
	}

	if len(data) != payloadSize {
		logger.Warn("deserialize-message-size", logger.Fail)
		return BodyMessage{}, errors.New("incomplete or oversized message")
	}

	offset := 0
	body := &BodyMessage{
		Dni: binary.BigEndian.Uint32(data[offset : offset+4]),
	}
	offset += 4

	body.Bet = binary.BigEndian.Uint32(data[offset : offset+4])
	offset += 4

	body.Year = binary.BigEndian.Uint16(data[offset : offset+2])
	offset += 2

	body.Month = data[offset]
	offset++

	body.Day = data[offset]
	offset++

	body.Name = string(data[offset : offset+nameSize])
	offset += nameSize

	body.SurName = string(data[offset : offset+surnameSize])
	return *body, nil
}

func Deserialize(data []byte) (Message, error) {
	header, err := deserialize_header(data)
	if err != nil {
		return Message{}, err
	}

	if header.SizePayload == 0 {
		return Message{
			Header: header,
			Body:   nil,
		}, nil
	}

	body, err := deserialize_body(data[MESSAGE_HEADER_SIZE:], header)
	if err != nil {
		return Message{}, err
	}
	return Message{
		Header: header,
		Body:   &body,
	}, nil
}

File: services/server/src/messages/messages.py

from dataclasses import dataclass
import socket
import string
import safe_socket
from lottery.bet import Bet

BODY_MIN_SIZE = 12
MESSAGE_HEADER_SIZE = 11
BATCH_HEADER_SIZE = 3

NAME_FIELD      = 0
SURNAME_FIELD   = 1
DNI_FIELD       = 2
DATE_FIELD      = 3
BET_FIELD       = 4

CONNECT          = 1
CONNECT_ACK      = 129
BATCH            = 16
BATCH_ACK        = 144
BET              = 64
BET_ACK          = 192
BET_END          = 66
WINNER           = 32
WINNER_ACK       = 160
CONNECT_END      = 3
CONNECT_END_ACK  = 131
ERROR            = 255

@dataclass
class HeaderMessage:
	Type:        int
	SeqNum:      int
	AckNum:      int
	AgencyId:    int
	SizePayload: int
	SizeName:    int
	SizeSurName: int

@dataclass
class BodyMessage:
	Dni:     int
	Bet:     int
	Year:    int
	Month:   int
	Day:     int
	Name:    str # type: ignore # max 50 bytes
	SurName: str # type: ignore # max 50 bytes

@dataclass
class Message:
	Header: HeaderMessage
	Body:   BodyMessage | None

@dataclass
class HeaderBatch:
	Type: int
	SizeBatch: int

@dataclass
class Batch:
	Header: HeaderBatch
	Messages: list[Message]

def read(socket) -> Message | Batch:
	header_data = safe_socket.recv_all(socket, BATCH_HEADER_SIZE)

	if header_data[0] == BATCH:
		header = deserialize_batch_header(header_data)
		if header.SizeBatch == 0:
			raise ValueError("batch cannot be empty")

		messages_in_batch = []
		for _ in range(header.SizeBatch):
			messages_in_batch.append(read_message(socket))

		return Batch(
			Header=header,
			Messages=messages_in_batch,
		)

	return read_message(socket, header_data)

def read_message(socket, header_data: bytes = b"") -> Message:
	if len(header_data) > MESSAGE_HEADER_SIZE:
		raise ValueError("message header prefix is too large")

	header_data += safe_socket.recv_all(socket, MESSAGE_HEADER_SIZE - len(header_data))
	header = deserialize_head(header_data)
	body = None

	if header.SizePayload > 0:
		body_data = safe_socket.recv_all(socket, header.SizePayload)
		body = deserialize_body(body_data, header)

	return Message(
		Header=header,
		Body=body,
	)

def send(socket: socket.socket, message: Message):
	batch = build_batch([message])
	binary_batch = serialize_batch(batch)
	safe_socket.send_all(socket, binary_batch)

def build_batch(messages_in_batch: list[Message]) -> Batch:
	return Batch(
		Header=HeaderBatch(
			Type=BATCH,
			SizeBatch=len(messages_in_batch),
		),
		Messages=messages_in_batch,
	)

def serialize_batch(batch: Batch) -> bytes:
	if batch.Header.SizeBatch != len(batch.Messages):
		raise ValueError("batch size does not match message count")

	if batch.Header.SizeBatch > 0xFFFF:
		raise ValueError("batch contains too many messages")

	data = bytearray(BATCH_HEADER_SIZE)
	data[0] = batch.Header.Type
	data[1:3] = batch.Header.SizeBatch.to_bytes(2, byteorder="big")

	for message in batch.Messages:
		data.extend(serialize(message))

	return bytes(data)

def deserialize_batch_header(data: bytes) -> HeaderBatch:
	if len(data) != BATCH_HEADER_SIZE:
		raise ValueError("invalid batch header size")

	if data[0] != BATCH:
		raise ValueError("invalid batch message type")

	return HeaderBatch(
		Type=data[0],
		SizeBatch=int.from_bytes(data[1:3], byteorder="big"),
	)
	

def build_message(kind: int, seq: int, ack: int, id: int, body: BodyMessage | None) -> Message:
	payload_size = 0
	name_size = 0
	surname_size = 0
	
	if body is not None:
		name_size = len(body.Name.encode("utf-8"))
		surname_size = len(body.SurName.encode("utf-8"))
		payload_size = BODY_MIN_SIZE + name_size + surname_size
	
	header = HeaderMessage(
		Type=kind,
		SeqNum=seq,
		AckNum=ack,
		AgencyId=id,
		SizePayload=payload_size,
		SizeName=name_size,
		SizeSurName=surname_size,
	)

	return Message(
		Header=header,
		Body=body,
	)
		
def message_to_bet(message: Message) -> Bet:
	if message.Header.Type != BET:
		raise ValueError("message is not a BET message")

	if message.Body is None:
		raise ValueError("BET message has no body")

	body = message.Body

	return Bet(
		agency_id=message.Header.AgencyId,
		first_name=body.Name,
		last_name=body.SurName,
		document=body.Dni,
		birthdate=f"{body.Year:04d}-{body.Month:02d}-{body.Day:02d}",
		number=body.Bet,
	)

def serialize(message: Message) -> bytes:
	data = serialize_head(message.Header)

	if message.Body is not None:
		data += serialize_body(message.Body)

	return data

def serialize_head(header: HeaderMessage) -> bytes:
	data = bytearray(MESSAGE_HEADER_SIZE)

	data[0] = header.Type
	data[1] = header.SeqNum
	data[2] = header.AckNum
	data[3:7] = header.AgencyId.to_bytes(4, byteorder="big")
	data[7:9] = header.SizePayload.to_bytes(2, byteorder="big")
	data[9] = header.SizeName
	data[10] = header.SizeSurName

	return bytes(data)

def serialize_body(body: BodyMessage) -> bytes:
	name = body.Name.encode("utf-8")
	surname = body.SurName.encode("utf-8")

	data = bytearray(BODY_MIN_SIZE + len(name) + len(surname))
	offset = 0

	data[offset:offset + 4] = body.Dni.to_bytes(4, byteorder="big")
	offset += 4

	data[offset:offset + 4] = body.Bet.to_bytes(4, byteorder="big")
	offset += 4

	data[offset:offset + 2] = body.Year.to_bytes(2, byteorder="big")
	offset += 2

	data[offset] = body.Month
	offset += 1

	data[offset] = body.Day
	offset += 1

	data[offset:offset + len(name)] = name
	offset += len(name)

	data[offset:offset + len(surname)] = surname

	return bytes(data)

def deserialize(data: bytes) -> Message:
	header = deserialize_head(data)

	if header.SizePayload == 0:
		return Message(
			Header=header,
			Body=None,
		)

	body = deserialize_body(data[MESSAGE_HEADER_SIZE:], header)

	return Message(
		Header=header,
		Body=body,
	)

def deserialize_head(data: bytes) -> HeaderMessage:
	if len(data) < MESSAGE_HEADER_SIZE:
		raise ValueError("header too small")
	
	return HeaderMessage(
		Type=data[0],
		SeqNum=data[1],
		AckNum=data[2],
		AgencyId=int.from_bytes(data[3:7], byteorder="big"),
		SizePayload=int.from_bytes(data[7:9], byteorder="big"),
		SizeName=data[9],
		SizeSurName=data[10],
	)

def deserialize_body(data: bytes, header: HeaderMessage) -> BodyMessage:
	payload_size = header.SizePayload
	name_size = header.SizeName
	surname_size = header.SizeSurName

	if payload_size < BODY_MIN_SIZE:
		raise ValueError("payload size is too short")

	if payload_size != BODY_MIN_SIZE + name_size + surname_size:
		raise ValueError("payload does not match with name sizes")

	if len(data) != payload_size:
		raise ValueError("message size does not match payload size")

	offset = 0
	dni = int.from_bytes(data[offset:offset+4], byteorder="big")
	offset += 4

	bet = int.from_bytes(data[offset:offset+4], byteorder="big")
	offset += 4

	year = int.from_bytes(data[offset:offset+2], byteorder="big")
	offset += 2

	month = data[offset]
	offset += 1

	day = data[offset]
	offset += 1

	name = data[offset:offset + name_size].decode("utf-8")
	offset += name_size

	surname = data[offset:offset + surname_size].decode("utf-8")

	return BodyMessage(
		Dni=dni,
		Bet=bet,
		Year=year,
		Month=month,
		Day=day,
		Name=name,
		SurName=surname,
	)
